### PR TITLE
[SPARK-47409][SQL] Add support for collation for StringTrim type of functions/expressions (for UTF8_BINARY & LCASE)

### PR DESCRIPTION
**Pull request by davidm-db**
Recreating [original PR](https://github.com/apache/spark/pull/45749) because code has been reorganized in [this PR](https://github.com/apache/spark/pull/45978).

### What changes were proposed in this pull request?
This PR is created to add support for collations to StringTrim family of functions/expressions, specifically:
- `StringTrim`
- `StringTrimBoth`
- `StringTrimLeft`
- `StringTrimRight`

Changes:
- `CollationSupport.java`
  - Add new `StringTrim`, `StringTrimLeft` and `StringTrimRight` classes with corresponding logic.
  - `CollationAwareUTF8String` - add new `trim`, `trimLeft` and `trimRight` methods that actually implement trim logic.
- `UTF8String.java` - expose some of the methods publicly.
- `stringExpressions.scala`
  - Change input types.
  - Change eval and code gen logic.
- `CollationTypeCasts.scala` - add `StringTrim*` expressions to `CollationTypeCasts` rules.

### Why are the changes needed?
We are incrementally adding collation support to a built-in string functions in Spark.

### Does this PR introduce _any_ user-facing change?
Yes:
- User should now be able to use non-default collations in string trim functions.

### How was this patch tested?
Already existing tests + new unit/e2e tests.

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CollationTypeCasts.scala [hunk 1]
```diff
@@ -18,11 +18,9 @@
 package org.apache.spark.sql.catalyst.analysis
 
 import javax.annotation.Nullable
-
```
Comment (mihailom-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CollationTypeCasts.scala:21]: Please remove unnecessary changes of line deletion.

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java [hunk 2]
```diff
@@ -214,6 +214,207 @@ public static int execICU(final UTF8String string, final UTF8String substring,
...
+    public static UTF8String exec(
+        final UTF8String srcString,
+        final UTF8String trimString,
+        final int collationId) {
+      CollationFactory.Collation collation = CollationFactory.fetchCollation(collationId);
+      if (collation.supportsBinaryEquality) {
+        return execBinary(srcString, trimString);
+      } else {
+        return execICU(srcString, trimString, collationId);
+      }
```
Comment (uros-db) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java:237]: why do we not have an `lcase` implementation?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CollationTypeCasts.scala [hunk 3]
```diff
@@ -54,7 +52,8 @@ object CollationTypeCasts extends TypeCoercionRule {
 
     case otherExpr @ (
       _: In | _: InSubquery | _: CreateArray | _: ArrayJoin | _: Concat | _: Greatest | _: Least |
-      _: Coalesce | _: BinaryExpression | _: ConcatWs) =>
+      _: Coalesce | _: BinaryExpression | _: ConcatWs | _: StringTrim | _: StringTrimBoth |
+      _: StringTrimLeft | _: StringTrimRight) =>
```
Comment (mihailom-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CollationTypeCasts.scala:56]: StringTrimBoth is replaced with StringTrim, so it is unnecessary to add it to rules here.

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java [hunk 4]
```diff
@@ -214,6 +214,207 @@ public static int execICU(final UTF8String string, final UTF8String substring,
...
+  public static class StringTrimLeft {
+    public static UTF8String exec(
+        final UTF8String srcString,
+        final int collationId) {
+      CollationFactory.Collation collation = CollationFactory.fetchCollation(collationId);
+      if (collation.supportsBinaryEquality) {
+        return execBinary(srcString);
+      } else {
+        return execICU(srcString, collationId);
+      }
```
Comment (uros-db) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java:293]: why do we not have an `lcase` implementation?

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java
Comment (uros-db) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java:1]: this doesn't look good

please follow the guides set out in https://issues.apache.org/jira/browse/SPARK-47410 regarding the latest refactor, and try to follow the structure of other string expressions (i.e. Contains, StartsWith, EndsWith)

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java
Comment (uros-db) [inline common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java:1]: I know that we're trying to push all collation aware changes away from UTF8String, and into CollationSupport.CollationAwareUTF8String

but we should be careful about changing access modifiers here, adding @cloud-fan to take a look and advise on what's the preferred approach:
1. modify access in UTF8String in order to allow collation-aware implementation in CollationSupport
2. fall back to putting everything in UTF8String, instead of using CollationSupport.CollationAwareUTF8String

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java
Comment (uros-db) [reply common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java:1]: essentially, we only want 1 `exec` (and 1 `genCode`), which should take a collationId, and then cover all 3 cases (execBinary, execLowercase, execICU)

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala [hunk 5]
```diff
@@ -1150,12 +1166,11 @@ case class StringTrim(srcStr: Expression, trimStr: Option[Expression] = None)
...
   override protected def direction: String = "BOTH"
 
-  override def doEval(srcString: UTF8String): UTF8String = srcString.trim()
+  override def doEval(srcString: UTF8String): UTF8String =
+    CollationSupport.StringTrim.exec(srcString, collationId)
 
   override def doEval(srcString: UTF8String, trimString: UTF8String): UTF8String =
-    srcString.trim(trimString)
-
-  override val trimMethod: String = "trim"
```
Comment (uros-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:1158]: why are you removing this?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala [hunk 6]
```diff
@@ -1258,12 +1273,11 @@ case class StringTrimLeft(srcStr: Expression, trimStr: Option[Expression] = None
...
   override protected def direction: String = "LEADING"
 
-  override def doEval(srcString: UTF8String): UTF8String = srcString.trimLeft()
+  override def doEval(srcString: UTF8String): UTF8String =
+    CollationSupport.StringTrimLeft.exec(srcString, collationId)
 
   override def doEval(srcString: UTF8String, trimString: UTF8String): UTF8String =
-    srcString.trimLeft(trimString)
-
-  override val trimMethod: String = "trimLeft"
```
Comment (uros-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:1266]: why are you removing this?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala [hunk 7]
```diff
@@ -1319,12 +1333,11 @@ case class StringTrimRight(srcStr: Expression, trimStr: Option[Expression] = Non
...
   override protected def direction: String = "TRAILING"
 
-  override def doEval(srcString: UTF8String): UTF8String = srcString.trimRight()
+  override def doEval(srcString: UTF8String): UTF8String =
+    CollationSupport.StringTrimRight.exec(srcString, collationId)
 
   override def doEval(srcString: UTF8String, trimString: UTF8String): UTF8String =
-    srcString.trimRight(trimString)
-
-  override val trimMethod: String = "trimRight"
```
Comment (uros-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:1327]: why are you removing this?

File: sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala [hunk 8]
```diff
@@ -536,6 +538,134 @@ class CollationStringExpressionsSuite
...
+          ")")
+      }
+      else {
+        // While other functions have arguments in (trimStr, srcStr) order
+        df = sql(s"SELECT ${testCase.trimFunc}(" +
+          (if (!testCase.hasTrimString) "" else if (testCase.trimString == null) "null, " else s"COLLATE('${testCase.trimString}', '${testCase.collation}'), ") +
+          s"COLLATE('${testCase.sourceString}', '${testCase.collation}')" +
+          ")")
+      }
+
```
Comment (uros-db) [inline sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala:662]: please take a look at some other tests in this suite (i.e. Contains/StartsWith/EndsWith)

if you made changes to collation type casts, then we need to test implicit casting here

File: sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala [hunk 9]
```diff
@@ -536,6 +538,134 @@ class CollationStringExpressionsSuite
...
+      checkAnswer(df = df, expectedAnswer = Row(testCase.expectedResultString))
+    })
+
```
Comment (uros-db) [inline sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala:665]: also, collation mismatch test is missing here

**Review by uros-db — CHANGES_REQUESTED** [review]
some changes needed

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CollationTypeCasts.scala (hunk 1 above)
Comment (davidm-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CollationTypeCasts.scala:21]: by accident, thanks!

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java (hunk 2 above)
Comment (davidm-db) [reply common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java:237]: You requested that we don't have `lcase` separately in the older PR: https://github.com/apache/spark/pull/45749#discussion_r1546312983.

The logic is now embeded in `CollationAwareUTF8String.trimLeft` and `CollationAwareUTF8String.trimRight`.

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java (hunk 4 above)
Comment (davidm-db) [reply common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java:293]: Reply in the other comment.

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java
Comment (davidm-db) [reply common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java:1]: Well, other methods like `numBytes` and `getBytes` are already exposed publicly. These methods seem as "important" and "risky" as the methods I changed. The methods I changed also don't modify anything, they are either helpers or they access some information that is already exposed through `numBytes` or `getBytes`.

That was my reasoning, but let's wait for @cloud-fan to provide his thoughts.

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java (hunk 2 above)
Comment (uros-db) [reply common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java:237]: that was indeed the case, but now it would be best if we could pull off separating these things

- static classes for expressions in CollationSupport should handle "rounting" of binary/lowecase/ICU implementations
- CollationAwareUTF8String should "extend" UTF8String with appropriate collation awareness, but shouldn't branch based on collation

for example, `CollationAwareUTF8String.trimLeft` should no longer call `CollationFactory.fetchCollation(collationId).supportsLowercaseEquality`, but execLowercase implementation in the appropriate `CollationSupport` class should do trimString.toLowerCase() prior to calling it

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java
Comment (davidm-db) [reply common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java:1]: Per my understanding, logic here is the same as in `Contains`, `StartsWith`, etc.

As I explained yesterday, `StringTrim*` functions work in specific way:
- If `trimString` param is not provided, then only spaces are trimmed.
- If `trimString` param is provided, then there are two options:
  - `trimString = null` - in this case `null` is returned always.
  - otherwise, characters from `trimString` are trimmed.

For this reason, in the original implementation, the methods were separated to the methods that have `trimString` param and to the methods that don't.
I only kept this reasoning for the sake of not changing default/expected behavior.

If you still think that we need to change to having only one method of each (`exec`, `genCode`, `execBinary` and `execICU`), let's sync offline to figure out how best to do this and have original logic perserved.

Comment (uros-db) [reply common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java:1]: hmm you are correct, this expression is a bit different than the ones above

we'll figure it out together offline what would be the best approach in this case

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala (hunk 5 above)
Comment (davidm-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:1158]: It was only used in `doGenCode` in the superclass.

After we changed to use `CollationSupport` classes/methods, and not call methods from `UTF8String` directly, there is no need for this field anymore.

Earlier this field was used in `doGenCode` to call specific method of `UTF8String`. Now, if we kept same behavior to call methods from `CollationSupport` we would either need to have a lot of hardcoded string compares or use reflection to invoke methods.

I think that this behavior where we match classes in `doGenCode` and then call respective methods from `CollationSupport` is more performant and robust.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala (hunk 6 above)
Comment (davidm-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:1266]: Replied in the other comment.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala (hunk 7 above)
Comment (davidm-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:1327]: Replied in the other comment.

File: sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala (hunk 8 above)
Comment (davidm-db) [reply sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala:662]: Yeah, I created initial set of tests now, but I'm not yet sure what exactly happened after Mihailo's changes regarding casting. I'll need to sync with him to understand better his changes before adding more related tests.

Earlier we had type mismatch checks in `stringExpressions`, but I see he has removed those, so I didn't understand that part yet. I'll sync with him and update the PR accordingly afterwards.

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java (hunk 2 above)
Comment (davidm-db) [reply common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java:237]: All right, will change accordingly and add separate methods to `CollationAwareUTF8String`.

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java
Comment (uros-db) [reply common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java:1]: I agree, these don't appear to be too dangerous, and I also think Milan needed some of these particular methods in his PRs as well (https://github.com/apache/spark/pull/45704), so it looks like we do need these to be public

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala (hunk 5 above)
Comment (uros-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:1158]: got it, sounds good

File: sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala (hunk 8 above)
Comment (uros-db) [reply sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala:662]: when you added `_: StringTrim | _: StringTrimLeft | _: StringTrimRight` to that last case in CollationTypeCasts, that should ensure that casting is enforced

so now just add tests here to confirm that it works

Comment (uros-db) [reply sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala:662]: like just modify the query to do `'${testCase.sourceString}'` instead of `COLLATE('${testCase.sourceString}', '${testCase.collation}')"` and it should work the same if COLLATE is called on the other parameter

File: common/unsafe/src/test/java/org/apache/spark/unsafe/types/CollationSupportSuite.java [hunk 10]
```diff
@@ -528,6 +528,235 @@ public void testFindInSet() throws SparkException {
...
+    assertStringTrim("UNICODE", "Ëaaaẞ", "Ëẞ", "aaa");
+    assertStringTrimLeft("UNICODE", "Ëaaaẞ", "Ëẞ", "aaaẞ");
+    assertStringTrimRight("UNICODE", "Ëaaaẞ", "Ëẞ", "Ëaaa");
+
+    assertStringTrim("UNICODE_CI", "Ëaaaẞ", "Ëẞ", "aaa");
+    assertStringTrimLeft("UNICODE_CI", "Ëaaaẞ", "Ëẞ", "aaaẞ");
+    assertStringTrimRight("UNICODE_CI", "Ëaaaẞ", "Ëẞ", "Ëaaa");
+
+    // weird i
+    // TODO: figure out why this doesn't work...
```
Comment (davidm-db) [inline common/unsafe/src/test/java/org/apache/spark/unsafe/types/CollationSupportSuite.java:755]: Need to figure out why these don't work properly.

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java
Comment (davidm-db) [reply common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java:1]: Decided on the approach offline and implemented it.

File: sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala (hunk 8 above)
Comment (davidm-db) [reply sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala:765]: Done.

File: sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala (hunk 9 above)
Comment (davidm-db) [reply sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala:768]: Added.

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java [hunk 11]
```diff
@@ -306,6 +308,258 @@ public static int execICU(final UTF8String string, final UTF8String substring,
...
+      return srcString.trim();
+    }
+    public static UTF8String execBinary(
+        final UTF8String srcString,
+        final UTF8String trimString) {
+      return srcString.trim(trimString);
+    }
+    public static UTF8String execLowercase(
+        final UTF8String srcString) {
+      return srcString.trim();
```
Comment (uros-db) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java:375]: this doesn't look very lowercase aware, what's the context?

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java [hunk 12]
```diff
@@ -403,6 +657,289 @@ private static int indexOf(final UTF8String target, final UTF8String pattern,
...
+        stringCharLen[numChars] = UTF8String.numBytesForFirstByte(srcString.getByte(byteIdx));
+        byteIdx += stringCharLen[numChars];
+        numChars++;
+      }
+
+      // Create ICU StringSearch object.
+      StringSearch stringSearch = CollationFactory.getStringSearch(
+        trimString, UTF8String.EMPTY_UTF8, collationId);
+      // Create hash set to save seen chars
+      Set<UTF8String> seenChars = new HashSet<>();
```
Comment (uros-db) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java:899]:
```suggestion
      Set<UTF8String> trimmedChars = new HashSet<>();
```

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java [hunk 13]
```diff
@@ -403,6 +657,289 @@ private static int indexOf(final UTF8String target, final UTF8String pattern,
...
+          continue;
+        }
+
+        // Otherwise, try to find the matching for the searchChar in the trimString.
+        stringSearch.reset();
+        stringSearch.setPattern(searchChar.toString());
+        int searchCharIdx = stringSearch.next();
+
+        if (searchCharIdx != StringSearch.DONE
+            && stringSearch.getMatchLength() == stringSearch.getPattern().length()) {
```
Comment (uros-db) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java:923]: here's the problem: we shouldn't check "stringSearch.getMatchLength() == stringSearch.getPattern().length()" because this is a false assumption (we are assuming that the match, which comes from target, and pattern will have the same number of characters - this is **not** true for "i̇" and "İ")

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java [hunk 14]
```diff
@@ -403,6 +657,289 @@ private static int indexOf(final UTF8String target, final UTF8String pattern,
...
+          trimByteIdx -= stringCharLen[numChars - 1];
```
Comment (uros-db) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java:924]: this is also a false assumption, we assume that we need to "move right by 1 character" - this is **not** true for "i̇" and "İ"

Comment (uros-db) [reply common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java:924]: instead, we should move right by `stringSearch.getMatchLength()` characters

consider the example of trying to do `trim("i̇abc", "İ")` so after we are done searching for "İ" in the target string, we need to trim `getMatchLength()` (which is 2) characters from "i̇abc", and then move on to trimming "abc" (which is a 2 character jump from the zeroth character)

now consider the example of trying to do `trim("İabc", "i̇")` so after we are done searching for "i̇" in the target string, we need to trim `getMatchLength()` (which is 1) characters from "İabc", and then move on to trimming "abc" (which is a 1 character jump from the zeroth character)

border line: always use `getMatchLength()` as the only valid source of truth, and don't try to count any bytes/chars manually, **especially** not using pattern.length()

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java (hunk 13 above)
Comment (uros-db) [reply common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java:923]: instead, we should assume that `stringSearch.getPattern().length()` will be different than `stringSearch.getMatchLength()` in some cases (this is fine, ICU took care of this and that's why StringSearch interface is the way it is)

consider the example of trying to do trim("i̇abc", "İ") so after we are done searching for "İ" in the target string, `getMatchLength()` will be 2 because it arises from the corresponding target substring (which is "i̇"), but pattern is still "İ" (which is a 1 character), hence `stringSearch.getMatchLength() != stringSearch.getPattern().length()
` (which is absolutely **fine**)

now consider the example of trying to do trim("İabc", "i̇") so after we are done searching for "i̇" in the target string, `getMatchLength()` will be 1 because it arises from the corresponding target substring (which is "İ"), but pattern is still "i̇" (which is a 2 characters), hence `stringSearch.getMatchLength() != stringSearch.getPattern().length()
` (which is absolutely **fine**)

border line: always use `getMatchLength()` as the only valid source of truth, and don't try to assume that you will know in advance how long the match is, especially not using pattern.length()

File: common/unsafe/src/test/java/org/apache/spark/unsafe/types/CollationSupportSuite.java (hunk 10 above)
Comment (uros-db) [reply common/unsafe/src/test/java/org/apache/spark/unsafe/types/CollationSupportSuite.java:761]: the idea is **not** to assume anything about match length, use `StringSearch` interface only (i.e. `getMatchLength`) to do **all** character operations, such as:

- moving on to the next character in the target sequence, when performing a trim
- storing processed target character in the hash set, when searching for matches
- doing the actual search using target & pattern

bear in mind, matches may actually be 2 characters in some special cases, even if pattern is only 1 char

Comment (uros-db) [reply common/unsafe/src/test/java/org/apache/spark/unsafe/types/CollationSupportSuite.java:761]: also: it would be useful to include the (wrong) test output of the current implementation, so that we could maybe understand better where it went wrong

**Review by uros-db — CHANGES_REQUESTED** [review]
great improvements across the board, even though there is a considerable amount of code in this PR, now everything is clearly separated, and more optimized too

we will however need a small fix for the ICU implementation, but otherwise looks great

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java (hunk 11 above)
Comment (mihailom-db) [reply common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java:375]: Shouldn't this be for a space character trim? ' ' is the same for lowercase?

File: sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala [hunk 15]
```diff
@@ -608,6 +610,181 @@ class CollationStringExpressionsSuite
...
+    checkEvaluation(
+      StringTrim(Literal.create("xxasdxx", StringType(0)), Literal.create("x", StringType(0))),
+      "asd")
+    checkEvaluation(
+      StringTrimLeft(Literal.create("xxasdxx", StringType(0)), Literal.create("x", StringType(0))),
+      "asdxx")
+    checkEvaluation(
+      StringTrimRight(Literal.create("xxasdxx", StringType(0)), Literal.create("x", StringType(0))),
+      "xxasd")
+  }
```
Comment (mihailom-db) [inline sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala:639]: I am not sure we need this test that much, when we have E2E, but if we intend to keep it, please use StringType(collation_name) as it makes it more readable for anyone who wants to review the code. Also extracting that name as a val collationName = "UTF8_BINARY" could be good.

File: common/unsafe/src/test/java/org/apache/spark/unsafe/types/CollationSupportSuite.java (hunk 10 above)
Comment (mihailom-db) [reply common/unsafe/src/test/java/org/apache/spark/unsafe/types/CollationSupportSuite.java:761]: I do not know if this helps, but this character behaves really weird. Uppercasing and then lowercasing it again does not produce the same character.

**Review by uros-db — CHANGES_REQUESTED** [review]
based on our offline discussion, ICU implementation proved to be a bit more complicated than expected, with lots of hard-to-catch edge cases

these are already pretty big changes from your side, so please remove the ICU implementation from this PR and we will update the ticket accordingly (only binary & lowercase support in this effort)

thanks David!

**Review by uros-db — CHANGES_REQUESTED** [review]
we also need tests for implicit collation in CollationStringExpressions.scala

btw the [ticket](https://issues.apache.org/jira/browse/SPARK-47409) is now updated to reflect that the scope is now **binary & lowercase collations only**

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationAwareUTF8String.java [hunk 16]
```diff
@@ -0,0 +1,304 @@
...
+      // Nothing trimmed.
+      return srcString;
+    }
+    if (trimByteIdx < 0) {
+      // Everything trimmed.
+      return UTF8String.EMPTY_UTF8;
+    }
+    return srcString.copyUTF8String(0, trimByteIdx);
+  }
+}
```
Comment (uros-db) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationAwareUTF8String.java:304]: add empty line at the end of file

Comment (davidm-db) [reply common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationAwareUTF8String.java]: Done.

**Review by uros-db — APPROVED** [review]
lgtm, thanks David for pushing this through!

@cloud-fan ready for review, also note: we separated out `CollationAwareUTF8String.java` in this PR

**Review by uros-db — CHANGES_REQUESTED** [review]
let's resolve conflicts and revive this

**Comment by uros-db** [comment]
@cloud-fan some tests are stale here, but overall I think the changes in this PR are good
however, changes are indeed large and merge conflicts arise daily - could you please review this?

**Review by cloud-fan — APPROVED** [review]

**Review by uros-db — APPROVED** [review]
I think we're ready to merge this

**Comment by cloud-fan** [comment]
thanks, merging to master!
